src/range_proof.rs

```diff
@@ -0,0 +1,555 @@
+                .enumerate()
+                .take(statement.generators.gi_base().len())
+            {
+                if gi_base[j] != *this_gi_base {
+                    return Err(ProofError::InternalDataInconsistent(
+                        "Inconsistent Gi generator point vector in batch statement".to_string(),
+                    ));
+                }
+            }
+            for (j, this_hi_base) in hi_base
+                .iter()
+                .enumerate()
+                .take(statement.generators.hi_base().len())
+            {
+                if hi_base[j] != *this_hi_base {
+                    return Err(ProofError::InternalDataInconsistent(
+                        "Inconsistent Hi generator point vector in batch statement".to_string(),
+                    ));
+                }
+            }
+        }
+
+        // Compute log2(N)
+        let mut log_n = 0;
+        let mut temp_n = bit_length >> 1;
+        while temp_n != 0 {
+            log_n += 1;
+            temp_n >>= 1;
+        }
+
+        // Compute 2**N-1 for later use
+        let mut two_n_minus_one = Scalar::from(2u8);
+        for _ in 0..log_n {
+            two_n_minus_one = two_n_minus_one * two_n_minus_one;
+        }
+        two_n_minus_one -= Scalar::one();
+
+        // Weighted coefficients for common generators
+        let mut g_base_scalar = Scalar::zero();
+        let mut h_base_scalar = Scalar::zero();
+        let mut gi_base_scalars = vec![Scalar::zero(); max_mn];
+        let mut hi_base_scalars = vec![Scalar::zero(); max_mn];
+
+        // Final multiscalar multiplication data
+        let mut scalars: Vec<Scalar> = vec![];
+        let mut points: Vec<RistrettoPoint> = vec![];
+
+        // Recovered masks
+        let mut masks = vec![];
+
+        // Process each proof and add it to the batch
+        let rng = &mut thread_rng();
+        for (index, proof) in range_proofs.iter().enumerate() {
+            let commitments = statements[index].commitments.clone();
+            let a = proof.get_a()?;
+            let a1 = proof.get_a1()?;
+            let b = proof.get_b()?;
+            let r1 = proof.get_r1();
+            let s1 = proof.get_s1();
+            let d1 = proof.get_d1();
+            let li = proof.get_li()?;
+            let ri = proof.get_ri()?;
+
+            if li.len() != ri.len() {
+                return Err(ProofError::InvalidLength(
+                    "Vector L length not equal to vector R length".to_string(),
+                ));
+            }
+            if 1 << li.len() != commitments.len() * bit_length {
+                return Err(ProofError::InvalidLength(
+                    "Vector L length not adequate".to_string(),
+                ));
+            }
+
+            // Helper values
+            let batch_size = commitments.len();
+            let rounds = li.len();
+
+            // Batch weight (may not be equal to a zero valued scalar)
+            let weight = Scalar::random(rng);
```

Comment:
Can't this function return zero? If so, need to check and loop until a nonzero scalar is returned.

Reply:
It is initialized with zero (`let mut scalar_bytes = [0u8; 64];`) and then filled with new random data.

Reply:
Right, but in theory it could return zero (or if the RNG were to fail somehow). It's cheap to check.

Reply:
O second thought, I could not find the check, so adding one.

Reply:
I did find a documentation reference that suggested the underlying byte generator _should_ not return the same output as the (zero) input, but adding a specific check here seems safer and more robust against implementation-specific behavior; it also makes the intent of the protocol more clear.